### PR TITLE
Send AuthFlow telemetry events to the App Insights

### PR DESCRIPTION
**Pull request by mvanchaa**
This is the first PR for the feature - Improving the AzureAuth telemetry instrumentation.

The final agenda is to make the following schema changes:

Drawing

Changes in this PR:
1. Added Lasso to MSALWrapper as we need `ITelemetryService` and `EventData` classes to send custom events of Auth Flows.
2. Changed `CommandMain` and `AuthFlowExecutor` classes to make use of `ITelemetryService`.
3. Populate and send `EventData` from `AuthFlowExecutor` class.

TODO:
1. Unit Tests (adding new and updating some old ones)
2. Add new attributes mentioned in the above schema to the main event `azureauth_command_azureauth`. I will create a new PR for this.
3. Fix Authentication type logic - (Currently, we capture authentication types as silent/interactive/devicecodeflow. These are not enough/accurate to capture whether a given authentication session was silent or not as the type - devicecodeflow can indicate a silent or interactive auth prompt. So there is a need for a flag to be added to the schema which indicates whether the authentication was silent or not.). I will create a new PR for this.

### CODE REVIEW

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs [hunk 1]
```diff
@@ -17,15 +19,18 @@ public class AuthFlowExecutor : IAuthFlow
...
         /// Initializes a new instance of the <see cref="AuthFlowExecutor"/> class.
         /// </summary>
         /// <param name="logger">The logger.</param>
+        /// <param name="telemetryService">The telemetry service.</param>
         /// <param name="authFlows">The list of auth flows.</param>
-        public AuthFlowExecutor(ILogger logger, IEnumerable<IAuthFlow> authFlows)
+        public AuthFlowExecutor(ILogger logger, ITelemetryService telemetryService, IEnumerable<IAuthFlow> authFlows)
         {
             this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.telemetryService = telemetryService;
```
Comment (kyle-rader) [inline src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:33]: Consider following the same pattern here that we expect non-null args and throw an exception if given a null arg.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs [hunk 2]
```diff
@@ -69,5 +76,11 @@ public async Task<AuthFlowResult> GetTokenAsync()
 
             return result;
         }
+
+        private void SendTelemetryEvent(AuthFlowResult attempt, string authFlowName)
+        {
+            // will be implemented in the next PR
+            return;
+        }
```
Comment (kyle-rader) [inline src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:84]: I'm not sure we need the helper method here yet. I'd rather see us collect the metrics about this auth flow and send it first before pre-abstracting. 

I think in this PR, it's fine to go ahead and actually send the event and will make the whole change easier to reason about seeing the full change in the same diff. 

To send an event, you need an event name, and an instance of `EventData`. Consider creating your EventData instance at the start of the for loop - one for each authFlow attempt. You can then add values to that eventData, and at the end of the loop, send the event constructing an event name based off the authFlowName.

**Review by kyle-rader — CHANGES_REQUESTED** [review]
Great, start - I don't think you need to break this change up and it will be easier to review the collection of the telemetry if we do adding the sending. Effectively this PR is 90% of the way to sending a new auth flow event, and doing the last 10% in a different PR makes the whole change harder to review.

File: src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs [hunk 3]
```diff
@@ -86,7 +86,7 @@ public void ConstructorWith_Null_AuthFlows()
         public void ConstructorWith_Valid_Arguments()
         {
             var logger = this.serviceProvider.GetService<ILogger<AuthFlowExecutor>>();
-            Action authFlowExecutor = () => new AuthFlowExecutor(logger, this.authFlows);
+            Action authFlowExecutor = () => new AuthFlowExecutor(logger, null, this.authFlows);
```
Comment (shalinikhare27) [inline src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs:89]: Shouldn't this be non-null for valid args?

File: src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs [hunk 4]
```diff
@@ -55,9 +55,9 @@ public void Setup()
         }
 
         [Test]
-        public void ConstructorWith_BothNullArgs()
+        public void ConstructorWith_AllNullArgs()
```
Comment (shalinikhare27) [inline src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs:58]: We probably need a test for valid telemetry service, but logger ad auth flows as null?

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 1 above)
Comment (shalinikhare27) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:33]: +1 to Kyle.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs [hunk 5]
```diff
@@ -48,9 +53,11 @@ public async Task<AuthFlowResult> GetTokenAsync()
                 this.logger.LogDebug($"Starting {authFlowName}...");
 
                 var attempt = await authFlow.GetTokenAsync();
+                this.SendTelemetryEvent(attempt, authFlowName);
+
                 if (attempt == null)
                 {
-                    var oopsMessage = $"Auth flow '{authFlow.GetType().Name}' returned a null AuthFlowResult.";
+                    var oopsMessage = $"Auth flow '{authFlowName}' returned a null AuthFlowResult.";
```
Comment (shalinikhare27) [inline src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:60]: How is the `authFlowName` set?

**Review by shalinikhare27 — CHANGES_REQUESTED** [review]
Waiting for some comment replies.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 2 above)
Comment (mvanchaa) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:84]: I thought of creating an instance of `EventData` inside each AuthFlow and then populating it inside the AuthFlow. When returning from any AuthFlow, this `EventData` instance will be returned along with the `AuthFlowResult`  like
```
return new AuthFlowResult(tokenResult, eventdata);
```

I will add all the changes inside one PR, I think that will make it easy to review.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 5 above)
Comment (mvanchaa) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:60]: This change is not highlighted because you added the same change I think and I merged your changes into this PR.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 1 above)
Comment (mvanchaa) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:33]: Got it!

File: src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs (hunk 3 above)
Comment (mvanchaa) [reply src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs:89]: Yup, that's true. I will change it.

File: src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs (hunk 4 above)
Comment (mvanchaa) [reply src/MSALWrapper.Test/AuthFlow/AuthFlowExecutorTest.cs:58]: Yes that's true.

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 2 above)
Comment (mvanchaa) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:93]: @kyle-rader add changes here. Let me know if this approach doesn't seem good.

File: src/MSALWrapper/AuthFlow/Broker.cs [hunk 6]
```diff
@@ -98,15 +108,18 @@ public async Task<AuthFlowResult> GetTokenAsync()
                             .GetTokenInteractiveAsync(this.scopes, account, cancellationToken),
                             this.errors)
                             .ConfigureAwait(false);
-                        tokenResult.SetAuthenticationType(AuthType.Interactive);
+                        tokenResult.SetAuthenticationType(AuthType.Interactive); // confirm this.
```
Comment (mvanchaa) [inline src/MSALWrapper/AuthFlow/Broker.cs:111]: @kyle-rader I was testing the code locally and I did not see an interactive prompt at this step. It seemed to have silently authenticate. Am I missing something here? 

I was testing on a machine with windows hello face enabled. I'm assuming it's because of that I couldn't see anything here. Else it would have prompted to take windows hello PIN. Am I right? 

My question here is, the authentication type could be `Interactive` or `Silent` depending on the machine configuration.  So which authentication type can be set here?

Also, I have added a new attribute - `no_of_interactive_prompts` to the schema. Whenever we invoke an interactive prompt, I'm incrementing the `interactivePromptsCount`. Am I right here?

File: src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs (hunk 2 above)
Comment (kyle-rader) [reply src/MSALWrapper/AuthFlow/AuthFlowExecutor.cs:93]: Generally I want to avoid having the auth logic itself know anything about Telemetry. I am _okay_ having the AuthFlowExector know about Telemetry given it's a meta class design to run multiple auth flows. But each auth flow I think should only return it's information, and the consumer can then decide what to send telemetry wise. The limits the responisbiltiy of every auth flow class, and forces knowledge about telemetry into 1 place. It also helps keep the telemetry schema per auth flow similar / the same.

**Review by kyle-rader — CHANGES_REQUESTED** [review]
See my main comment on not having any specific auth flow (except the auth flow executor, know about telemetry. Having each auth flow have to handle telemetry on it's own, as now seen, requires changing most files, and over-burdens the responsibility of each auth flow, + adds a lot of duplication that can be handled centrally by the auth flow executor. 

I was little on the fence about even adding the telemetry service to auth flow executor, but given it's current role, I think that's okay and makes sense, since it is the main orchestrator of auth flows.

File: src/MSALWrapper/ExceptionExtensions.cs [hunk 7]
```diff
@@ -66,4 +67,29 @@ public static IEnumerable<Exception> GetAllExceptions(this Exception exception)
             }
         }
     }
+
+    /// <summary>
+    /// Extracts correlation IDs from the list of exceptions.
+    /// </summary>
+    /// <param name="exceptions">List of exceptions from which correlation IDs are extracted.</param>
+    /// <returns>List of correlation IDs.</returns>
+    public static List<string> ExtractCorrelationIDsFromException(IList<Exception> exceptions)
```
Comment (kyle-rader) [inline src/MSALWrapper/ExceptionExtensions.cs:76]: Discussed offline - move this method into ExceptionListToStringConverter since it's not a true _extension_ method to the Exception class.

File: src/MSALWrapper/AuthFlow/Broker.cs [hunk 8]
```diff
@@ -23,6 +24,7 @@ public class Broker : IAuthFlow
         private readonly string promptHint;
         private readonly IList<Exception> errors;
         private IPCAWrapper pcaWrapper;
+        private int interactivePromptsCount;
```
Comment (kyle-rader) [inline src/MSALWrapper/AuthFlow/Broker.cs:27]: Discussed offline - because we do not have a specific question to answer by tracking prompt count, we don't need to collect this information right now. Generally speaking a run of an auth flow should have 0 (silent) or 1 (interactive) prompts and 1 is already more than desired. So we're going to remove this count.

**Review by kyle-rader — CHANGES_REQUESTED** [review]
Summary (from offline discussion):
Break apart the overal change into 3:
1. Add an immutable `AuthFlowName` property to `AuthFlowResult`
2. Change the interface of AuthFlowExecutor to return both the first successful `AuthFlowResult` as well as the full list of AuthFlowResults.
3. In CommandMain - utilize the now available data to send a separate event for each `AuthFlowResult` returned.
